src/jit/lsraarm64.cpp

```diff
@@ -971,103 +1006,22 @@ void Lowering::TreeNodeInfoInitCall(GenTreeCall* call)
-                {
-                    GenTreePtr putArgRegNode = fieldListPtr->Current();
-                    assert(putArgRegNode->gtOper == GT_PUTARG_REG);
-                    GenTreePtr putArgChild = putArgRegNode->gtOp.gtOp1;
-
-                    originalSize += REGSIZE_BYTES; // 8 bytes
-
-                    // Record the register requirements for the GT_PUTARG_REG node
-                    putArgRegNode->gtLsraInfo.setDstCandidates(l, targetMask);
-                    putArgRegNode->gtLsraInfo.setSrcCandidates(l, targetMask);
-
-                    // To avoid redundant moves, request that the argument child tree be
-                    // computed in the register in which the argument is passed to the call.
-                    putArgChild->gtLsraInfo.setSrcCandidates(l, targetMask);
-
-                    // We consume one source for each item in this list
-                    info->srcCount++;
-                    iterationNum++;
-
-                    // Update targetReg and targetMask for the next putarg_reg (if any)
-                    targetReg  = genRegArgNext(targetReg);
-                    targetMask = genRegMask(targetReg);
-                }
-            }
-            else
-            {
-#ifdef DEBUG
-                compiler->gtDispTreeRange(BlockRange(), argNode);
-#endif
-                noway_assert(!"Unsupported TYP_STRUCT arg kind");
-            }
-
-            unsigned  slots          = ((unsigned)(roundUp(originalSize, REGSIZE_BYTES))) / REGSIZE_BYTES;
-            regNumber curReg         = argReg;
-            regNumber lastReg        = argIsFloat ? REG_ARG_FP_LAST : REG_ARG_LAST;
-            unsigned  remainingSlots = slots;
-
-            while (remainingSlots > 0)
+            // There could be up to 2-4 PUTARG_REGs in the list (3 or 4 can only occur for HFAs)
+            regNumber argReg = curArgTabEntry->regNum;
+            for (GenTreeFieldList* entry = argNode->AsFieldList(); entry != nullptr; entry = entry->Rest())
             {
-                argMask |= genRegMask(curReg);
-                remainingSlots--;
+                TreeNodeInfoInitPutArgReg(entry->Current()->AsUnOp(), argReg, *info, false, &callHasFloatRegArgs);
 
-                if (curReg == lastReg)
-                    break;
-
-                curReg = genRegArgNext(curReg);
+                // Update argReg for the next putarg_reg (if any)
+                argReg = genRegArgNext(argReg);
             }
-
-            // Struct typed arguments must be fully passed in registers (Reg/Stk split not allowed)
-            noway_assert(remainingSlots == 0);
-            argNode->gtLsraInfo.internalIntCount = 0;
         }
-        else // A scalar argument (not a struct)
+        else
         {
-            // We consume one source
-            info->srcCount++;
-
-            argMask |= genRegMask(argReg);
-            argNode->gtLsraInfo.setDstCandidates(l, argMask);
-            argNode->gtLsraInfo.setSrcCandidates(l, argMask);
```

Comment:
Prior to this change `argMask` was guaranteed to be `RBM_NONE` at the point at which `genRegMask(argReg)` was or'd in (i.e. line 1059), so it would also have only contained the current arg reg mask. The behavior is the same in both cases.